### ISSUE TITLE
Spinner should only be visible once when multiple http requests are sent at almost the same time

### ISSUE THREAD

**Issue by whizkidwwe1217** [mpalourdio/ng-http-loader#85]
Use case:
I have two http requests called simultaneously:
1. The first http request is sent, the spinner is visible.
2. After 50ms, the second http request is sent. **Issue**: The spinner shows up again and since the time the spinner is shown and hidden between the two requests is instantaneous, it looks like it's flickering.

**Comment by mpalourdio** [comment]
This case is handled in this module. But, if you observe flickering between requests, it's because the first one ends **before** the second ones begins. There's not much we can do. Things you can try :

- perform your requests with **forkJoin** if it's possible
- use the **minDuration** option

**Comment by mpalourdio** [comment]
Closing for now. Feel free to reopen with a runnable sample that demonstrates the problem, or even better by submitting a PR with a failing test case.

**Comment by mpalourdio** [comment]
@gnom7  Can you please share a code snippet to reproduce this issue ?

**Comment by gnom7** [comment]
@mpalourdio This issue appeared in project from my work, where I have forced to downgrade to the version 2.0.0, so I cannot share source code of the project, but I will try to find time on weekend to create simple sample to reproduce the issue.

**Comment by mpalourdio** [comment]
ok ! Or feel free to PR a failing unit test.

**Comment by mpalourdio** [comment]
I just made some tests and cannot see any of those glitches. Which spinner do you use ?

**Comment by gnom7** [comment]
@mpalourdio I use `skThreeBounce`

**Comment by gnom7** [comment]
@mpalourdio 
I tried to reproduce this issue today with a new empty project, I finished up with [this](https://github.com/gnom7/ng-http-loader-blink-issue) project. Run `node server.js` and `ng serve`, then open `localhost:4200`. Most times you will see one wave of three-bounce spinner. Then you may change version from `2.0.0` to `2.1.0` and after this change all times you will see two half-waves of three-bounce spinner.
Not sure if the reason is the same on my work project, but it seems to be right -- we send several requests on some pages and some requests needs info from previous; so steps to repro would be
1) send first request
2) after it wait approximately 1ms (to simulate crafting of new request)
3) send second request

I also tried to use `minDuration` for latest version, but it helps only when it covers all (two in my sample) requests. I mean that in sample project I have two requests for page each ~300ms, setting `minDuration` to 400ms doesn't help, setting to 700ms helps, but then a lot of pages would seem to load much more time than they really are.
Maybe, to mitigate this issue, we can even return not `timer(0)` but `timer(5)` (I tested it and wasn't able to reproduce the issue not 'most times' but at all)? We can externalize it just like other properties. I mean can we introduce something like `minExtraDuration` - how long spinner should be shown after request ended? Setting this property to about 5-10ms will allow to get rid of blinks.
What would you say?

**Comment by mpalourdio** [comment]
Thanks A LOT for this repo. I was able to see what you are experiencing. But I'd like to point some things :

- I clearly can experience the flickering effect with version 2.10
- You do NOT run parallel request, their are performed synchronously. You can see it in the chrome console.
- The `minExtraDuration` option yet exists. It's name `minDuration` and was introduced in [2.1.0](https://github.com/mpalourdio/ng-http-loader/blob/master/CHANGELOG.md#v210). I've just tried it on your project by setting it to `1000` and fixes the behavior you are seeing. It had clearly been introduced for such cases.
- What you are experiencing is a combination + synchronous HTTP request + `debouceDelay`. This flickering makes me think that the module behaves correctly.

Now, I can still give it a look at why `timer(0)` performs differently than `EMPTY`, but clearly, as your HTTP requests run one after the other + the deBouceDelay the flickering looks okish  to me.

Just give V3 a try + `minDuration`, you'll see that all is ok ;)

Please give me feedback. And I really appreciate that you took time to allox me to reproduce the repo !

**Comment by gnom7** [comment]
@mpalourdio 
I'm glad to help you.
But still, have you tried setting `minDuration` to the 400ms for example? setting it to 1s seems a lot and won't work for multiple requests running in series more than one second. I'd expect the `minExtraDuration` to prolongate spinner visibility after request finished, `minDuration` counts from very beginning.

**Comment by mpalourdio** [comment]
Yes it does, because your HTTP requests, once again, are performed `synchronously`, and not in parallel. There's really nothing we can do here at the moment.

**Comment by gnom7** [comment]
@mpalourdio 
I have removed `debounceDelay` property from my sample -- the issue still there with version `2.1.0`.
I clearly realize that request are run synchronously, but that's my use case.
Then, changing `time(0)` to `time(5)` clearly fixes the issue.
I tried to set `minDuration` on my work project up to 7s increasing it step by step and it didn't help (I upgraded `angular` version from `5` to `6` and I have upgraded `ng-http-loader` directly to `3.0.0` and already tried to solve the issue with help of `minDuration`). I will try it once more, anyway. Am I right that `minDuration` starts counting from beginning of http request? Wouldn't it be better to count from end of request to better and clearer cover such cases as in sample project?

**Comment by mpalourdio** [comment]
Yes, I have just wrote a unit test for this case, and it clearly does not handle your particular use-case ! Changing the` timer(0)` to `timer(5)` is a no-go. I want to keep `EMPTY` as it is

But handling the `minDuration` for `synchronous` HTTP request is clearly something I can try to manage. But it's a feature request, not a bug fix. I still consider the observed behavior as expected. 

Can you please open a new issue for this case, so I can close this one and I'll we flag the new one as feature request ?

Thanks again for feedback !

**Comment by gnom7** [comment]
@mpalourdio 
> If minDuration is set to 400, and your first request takes 300ms, the second one 500, the spinner will hide by itself at 800ms (300+500), whatever is the minDuration. So I can't see how it can seem to load much more than it actually does TBH.

And if I set `minDuration` to 1s, then even with sole request which takes 300ms spinner will be shown as for 1s request, am I missing something?

**Comment by mpalourdio** [comment]
Yes, with a unique HTTP request, that's the current behavior. See this pseudo-marble https://github.com/mpalourdio/ng-http-loader/blob/master/CHANGELOG.md#v210

**Comment by gnom7** [comment]
@mpalourdio Thank you for the link, I already read this release notes, but then what would you say about different handling for `minDuration` as described above? Is this how you intend to deal with feature request issue I have just opened?
I'll be glad to have the requested feature, thank you for your help and this awesome library!

**Comment by mpalourdio** [comment]
Have you watched the commit of my unit test in your freshly opened issue ? Hopes it is what you expect. Let's continue discussion on your issued, as this one is closed and not related.
